homeassistant/components/lgthinq/__init__.py

```diff
@@ -0,0 +1,622 @@
+    )
+    if soundbar_client is None:
+        raise ConfigEntryError
+
+    hass.async_add_executor_job(soundbar_client.listen)
+    # runtime_data for soundbar
+    entry.runtime_data.soundbar_client = soundbar_client
+    await hass.config_entries.async_forward_entry_setups(entry, SOUNDBAR_PLATFORMS)
+    return True
+
+
+async def async_unload_entry_soundbar(
+    hass: HomeAssistant, entry: ThinqConfigEntry
+) -> bool:
+    """Unload a config entry."""
+    return await hass.config_entries.async_unload_platforms(entry, SOUNDBAR_PLATFORMS)
+
+
+##### webOS TV #####
+
+WEBOS_PLATFORMS = [Platform.MEDIA_PLAYER]
+CONFIG_SCHEMA = cv.removed(DOMAIN, raise_if_present=False)
+
+CALL_SCHEMA = vol.Schema({vol.Required(ATTR_ENTITY_ID): cv.comp_entity_ids})
+
+BUTTON_SCHEMA = CALL_SCHEMA.extend({vol.Required(ATTR_BUTTON): cv.string})
+
+COMMAND_SCHEMA = CALL_SCHEMA.extend(
+    {vol.Required(ATTR_COMMAND): cv.string, vol.Optional(ATTR_PAYLOAD): dict}
+)
+
+SOUND_OUTPUT_SCHEMA = CALL_SCHEMA.extend({vol.Required(ATTR_SOUND_OUTPUT): cv.string})
+
+SERVICE_TO_METHOD = {
+    SERVICE_BUTTON: {"method": "async_button", "schema": BUTTON_SCHEMA},
+    SERVICE_COMMAND: {"method": "async_command", "schema": COMMAND_SCHEMA},
+    SERVICE_SELECT_SOUND_OUTPUT: {
+        "method": "async_select_sound_output",
+        "schema": SOUND_OUTPUT_SCHEMA,
+    },
+}
+
+
+async def async_setup_webostv(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the LG WebOS TV platform."""
+    _LOGGER.warning("async_setup_webostv. config: %s", config)
+    hass.data[DOMAIN].setdefault(DATA_CONFIG_ENTRY_WEBOSTV, {})
+    hass.data[DOMAIN][DATA_HASS_CONFIG] = config
+
+    return True
+
+
+async def async_setup_entry_webostv(
+    hass: HomeAssistant, entry: ThinqConfigEntry
+) -> bool:
+    """Set up LG WebOS TV platform config entry."""
+    hass.data[DOMAIN].setdefault(DATA_CONFIG_ENTRY_WEBOSTV, {})
+
+    host = entry.data[CONF_HOST]
+    key = entry.data[CONF_CLIENT_SECRET]
+
+    # Attempt a connection, but fail gracefully if tv is off for example.
+    client = WebOsClient(host, key)
+    with suppress(*WEBOSTV_EXCEPTIONS):
+        try:
+            await client.connect()
+        except WebOsTvPairError as err:
+            raise ConfigEntryError(err) from err
+
+    # If pairing request accepted there will be no error
+    # Update the stored key
+    update_client_key(hass, entry, client)
+
+    async def async_service_handler(service: ServiceCall) -> None:
+        method = SERVICE_TO_METHOD[service.service]
+        data = service.data.copy()
+        data["method"] = method["method"]
+        async_dispatcher_send(hass, DOMAIN, data)
+
+    for service, method in SERVICE_TO_METHOD.items():
+        schema = method["schema"]
+        hass.services.async_register(
+            DOMAIN, service, async_service_handler, schema=schema
+        )
+
+    hass.data[DOMAIN][DATA_CONFIG_ENTRY_WEBOSTV][entry.entry_id] = client
```

Comment:
Please don't use hass.data for this anymore. Instead, store the runtime data in the config entry. For more information see:

https://developers.home-assistant.io/blog/2024/04/30/store-runtime-data-inside-config-entry